sql: add create table test

BramGruneir opened the pull request:
sql: add create table test
Closes #9012.

Also 2 small spelling fixes commits.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9182)

<!-- Reviewable:end -->

BramGruneir committed:
sql: add create table test

Closes #9012.

BramGruneir committed:
kv: spelling fix

BramGruneir committed:
sql: fix spelling errors

vivekmenezes commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/create_test.go, line 185 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4VrBpIYukMqADxxpR:-KR4VrBpIYukMqADxxpS:b32wo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L185)):*
> ```Go
> 
> 	createTestTable := func(
> 		ID int,
> ```

id

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KR4XUVTwD0YSGAZJ0hf:bcwbnz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
> ```Go
> 			completed.Lock()
> 			defer completed.Unlock()
> 			completed.tables[ID] = struct{}{}
> ```

I suppose you could add the ids to a channel and read the channel down below. That way you don't need the mutex

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 194 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4dgpSQzRQHOQb43ei:-KR4dgpSQzRQHOQb43ej:b-kmm29l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L194)):*
> ```Go
> 		select {
> 		case <-tc.Stopper().ShouldStop():
> 			t.Fatalf("text exiting, start signal never received")
> ```

no bueno, this is called from a goroutine other than the main one. why do you need this at all? elsewhere, we allow tests to time out.

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KR4dl4GpRzQRqgxZpyj:bgv3a98) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I suppose you could add the ids to a channel and read the channel down below. That way you don't need the mutex
</details>
+1 to channel based approach.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/create_test.go, line 185 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4VrBpIYukMqADxxpR:-KR4jOhedPUMzIXS9XVg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L185)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> id
</details>
Done.

---

*[sql/create_test.go, line 194 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4dgpSQzRQHOQb43ei:-KR4jRsnT-5mJR-w-KGM:b-y7zj4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L194)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> no bueno, this is called from a goroutine other than the main one. why do you need this at all? elsewhere, we allow tests to time out.
</details>
removed.  I had it in there when debugging the test.

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KR4pXVLsNAUXJicO2hx:bd9ccwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> +1 to channel based approach.
</details>
Done.  Much more elegant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[sql/create_test.go, line 184 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4odh0Y-RdFzR9n6o5:-KR4odh0Y-RdFzR9n6o6:b-dgz7f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L184)):*
> ```Go
> 	}
> 
> 	createTestTable := func(
> ```

the code for this test is big enough to where things are more readable if we just separate it into a "top level" function (same below)

---

*[sql/create_test.go, line 206 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4ozoHztclRWyAl3jN:-KR4ozoHztclRWyAl3jO:b-3l2bsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L206)):*
> ```Go
> 			if !hideErrors {
> 				t.Logf("table %d: could be created: %s", ID, err)
> 			}
> ```

[nit] can return here

---

*[sql/create_test.go, line 229 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4pzeLyfpimAKI1yrQ:-KR4pzeMtS_bABiq560P:b-6ghgam) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L229)):*
> ```Go
> 	for ID := range completed.tables {
> 		insertTableSQL := fmt.Sprintf(`
> 			INSERT INTO "test"."table_%d"
> ```

Is insertion necessary? (vs a simpler `SHOW CREATE TABLE`) 

---

*[sql/create_test.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qSN59EihqSeIvJg8:-KR4qSN59EihqSeIvJg9:b-ahy55s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L186)):*
> ```Go
> 	) {
> 		defer wg.Done()
> 		select {
> ```

this should just be `<-signal`

---

*[sql/create_test.go, line 236 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qfYMV8IKO6UJKV4Y:-KR4qfYMV8IKO6UJKV4Z:b-ecgq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L236)):*
> ```Go
> 
> 	completedTables1 := testAndCountTables(completed1)
> 	if completedTables1 < minimumCreatedTables {
> ```

Shouldn't all the creates succeed in this case? What failures are we seeing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 3 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 229 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4pzeLyfpimAKI1yrQ:-KR4t7lO3pd4D3fgY7ka:b-6pf4tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L229)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Is insertion necessary? (vs a simpler `SHOW CREATE TABLE`) 
</details>
I think it will be better to:
1. Ensure that the next id is set to the correct value: keys.MaxReservedDescID + numTablesCreated + 1
2. Read all the table descriptors using the expected ids. I think the range of ids will be [keys.MaxReservedDescID + 1, keys.MaxReservedDescID + numTablesCreated]

This then also doubles up as a test for serializability on increment :-)

---

*[sql/create_test.go, line 207 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4rOsnveQJ_vsdZrS2:-KR4rOsnveQJ_vsdZrS3:b-i6n09a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L207)):*
> ```Go
> 	wg1.Add(numberOfTables)
> 	signal1 := make(chan struct{})
> 	completed1 := make(chan int, numberOfTables)
> ```

why not reuse these vars instead of numbering them?

---

*[sql/create_test.go, line 215 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4sDk10DyUFqo96lbm:-KR4sDk2gtUfg8tuws88:bovpqvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L215)):*
> ```Go
> 	close(signal1)
> 	wg1.Wait()
> 	close(completed1)
> ```

do you really need to close the channel?

---

*[sql/create_test.go, line 221 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4sb5rvFj5cQp2z-WY:-KR4sb5swV7V75TdZDho:b-mz26fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L221)):*
> ```Go
> 	// were created.
> 	testAndCountTables := func(completed chan int) int {
> 		var numTablesCreated int
> ```

numTablesCreated := len(completed)

---

*[sql/create_test.go, line 236 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qfYMV8IKO6UJKV4Y:-KR4t4E2x1r4y3k4KYO2:bydoecg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L236)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Shouldn't all the creates succeed in this case? What failures are we seeing?
</details>
why not retry until success?

---

*[sql/create_test.go, line 259 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4uoC5dp3O0QB2qq7l:-KR4uoC5dp3O0QB2qq7m:b-t632rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L259)):*
> ```Go
> 	if completedTables2 != 1 {
> 		t.Fatalf("expected 1 table, got %d", completedTables2)
> 	}
> ```

same here. ensure that the id is only incremented once.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KR4wI-M_tIQ4Yq8xCMz:bmfz0nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.  Much more elegant.
</details>
Do you still need the waitgroup? I was imagining an error channel rather than a channel of successful IDs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I think it will also be valuable to clarify what exactly you are testing. It wouldn't be a bad idea to split the tests into two separate tests. Thanks!

---

Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KR4xuxFo-NrsAyNsbCk:ba05k5v)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Sure, I can easily split the test into two.

---

Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 184 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4odh0Y-RdFzR9n6o5:-KR506tllxEYus4omd_7:b-j5s057) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L184)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> the code for this test is big enough to where things are more readable if we just separate it into a "top level" function (same below)
</details>
Sure, can do.

---

*[sql/create_test.go, line 206 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4ozoHztclRWyAl3jN:-KR50Mflh_e0gY2rpxkH:b-64ihvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L206)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> [nit] can return here
</details>
Already so close to the end of the func already, it doesn't really save anything.

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KR4yP-5y1RKZEVQCdZg:b-kybyv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Do you still need the waitgroup? I was imagining an error channel rather than a channel of successful IDs.
</details>
The idea was to wait until all tables were created (or attempted at least). What would the waiting condition look like in that case?

---

*[sql/create_test.go, line 229 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4pzeLyfpimAKI1yrQ:-KR4zxGkDAx7CnlGC3qE:b-2eak6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L229)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I think it will be better to:
> 1. Ensure that the next id is set to the correct value: keys.MaxReservedDescID + numTablesCreated + 1
> 2. Read all the table descriptors using the expected ids. I think the range of ids will be [keys.MaxReservedDescID + 1, keys.MaxReservedDescID + numTablesCreated]
> 
> This then also doubles up as a test for serializability on increment :-)
</details>
Show would be fine, I just prefer performing a write as well. 

Interesting idea about ensuring that the ids are all there.  Let me think about that.

---

*[sql/create_test.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qSN59EihqSeIvJg8:-KR508fX4u4VCpD3m6P9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L186)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> this should just be `<-signal`
</details>
done

---

*[sql/create_test.go, line 207 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4rOsnveQJ_vsdZrS2:-KR4yqEW815sUuobd2Ys:bxrfgyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L207)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> why not reuse these vars instead of numbering them?
</details>
I just try not to reuse variables unless there is a good reason to.  Two different tests, so two different variables.  Anyway, split the test so it's not relevant anymore.

---

*[sql/create_test.go, line 215 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4sDk10DyUFqo96lbm:-KR4ywNcZaDT51WhGQTQ:bjk401y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L215)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> do you really need to close the channel?
</details>
the`for id := range completed {` won't end if the channel isn't closed.

---

*[sql/create_test.go, line 221 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4sb5rvFj5cQp2z-WY:-KR4y4Xe4FiYNbPeOO44:b76q8cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L221)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> numTablesCreated := len(completed)
</details>
I forgot you could do that.   Done.

---

*[sql/create_test.go, line 236 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qfYMV8IKO6UJKV4Y:-KR4zI4ED_ds03IyaNaZ:b-jmx9xr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L236)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> why not retry until success?
</details>
The goal is not to prove that we can insert n tables.  We can do that.  The goal of this test is to try to create n tables all at the same time. 

As far as failures are concerned, I get a lot of deadline exceeded  errors.  I'm going to spend some time investigating what the causes of these errors are.  Try setting numberOfTables to 100 or more and watch it either deadlock or fail.  I'll be filing an issue about this shortly.

---

*[sql/create_test.go, line 259 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4uoC5dp3O0QB2qq7l:-KR5-V2uzPgFIrE0vmN5:bebw1tn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L259)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> same here. ensure that the id is only incremented once.
</details>
Again, let me think about this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KR5-y3yH-TeYA4llylO:b-tn6bsb)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KR5AHXn8YgEyfT5b3KA:b-xdtt8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The idea was to wait until all tables were created (or attempted at least). What would the waiting condition look like in that case?
</details>
Reading an expected number of errors (some will be nil) from the channel.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 229 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4pzeLyfpimAKI1yrQ:-KR5AXe04zJs1SA20a2q:b80oxvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L229)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Show would be fine, I just prefer performing a write as well. 
> 
> Interesting idea about ensuring that the ids are all there.  Let me think about that.
</details>
I suspect this choice is going to follow from the intent of the test. I'd define that first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 3 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 229 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4pzeLyfpimAKI1yrQ:-KRUKfSuK4_wG65081i1:biew82u) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L229)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I suspect this choice is going to follow from the intent of the test. I'd define that first.
</details>
I also don't think this insertion is necessary. And also I don't know if we should test that the ids are consecutive... Doesn't seem like a guarantee the database should be making.

---

*[sql/create_test.go, line 157 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUGg4wOP6MUTcjRJsy:-KRUGg4wOP6MUTcjRJsz:bfe8ush) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L157)):*
> ```Go
> // TestCreateTable tests that concurrent create table requests are correctly
> // filled.
> func TestCreateTable(t *testing.T) {
> ```

put a "parallel" in the test name

---

*[sql/create_test.go, line 183 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUI-HbLBN2wRyfp7Zf:-KRUI-HclsLmupL3aHg6:b-jumbxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L183)):*
> ```Go
> 		signal chan struct{},
> 		completed chan int,
> 		hideErrors bool,
> ```

instead of a blanket `hideErrors`, pass in an expected error string.

---

*[sql/create_test.go, line 186 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qSN59EihqSeIvJg8:-KRUJ8VGZ1CEJtqTa1HP:b-hndcz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L186)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> done
</details>
not done?

---

*[sql/create_test.go, line 197 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUHA9eeib_2BC13yt2:-KRUHA9fdIA7OZAWDyYm:btluk6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L197)):*
> ```Go
> 		if _, err := db.Exec(tableSQL); err != nil {
> 			if !hideErrors {
> 				t.Logf("table %d: could be created: %s", id, err)
> ```

could *not* be created
What's this `t.Logf`? Haven't seen it before, just use `log.Infof()`? Or actually shouldn't this be `t.Errorf()` ?

---

*[sql/create_test.go, line 236 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qfYMV8IKO6UJKV4Y:-KRUJFiLIof5sQc9MPQg:b-d7f1p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L236)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The goal is not to prove that we can insert n tables.  We can do that.  The goal of this test is to try to create n tables all at the same time. 
> 
> As far as failures are concerned, I get a lot of deadline exceeded  errors.  I'm going to spend some time investigating what the causes of these errors are.  Try setting numberOfTables to 100 or more and watch it either deadlock or fail.  I'll be filing an issue about this shortly.
</details>
btw we got rid of the deadline on the distsender context, so results should now be different

---

*[sql/create_test.go, line 249 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUKFTJ9yyXmsXdmqd4:-KRUKFTJ9yyXmsXdmqd5:b-g9b1hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L249)):*
> ```Go
> 	for i := 0; i < numberOfTables; i++ {
> 		db := tc.ServerConn(i % numberOfNodes)
> 		go createTestTable(finalTableNumber, db, &wg2, signal2, completed2, true /*hideErrors*/)
> ```

the bug that triggered these tests would only be observed by this test, which actually has failing `CREATE`s, right? So you're using the duplica name just to force some errors, you're not actually interested in testing anything about duplicate names, are you? If you, there's a way to inject stmt errors which I think should be useful here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Bram I think you haven't pushed the latest.

---

Review status: 2 of 3 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRULeS5dDmnc_KYjjwD:b-mquexp)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I'm about to.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRUS5xNgRNNyuCPz5WA:bt3xaay)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 3 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[sql/create_test.go, line 211 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4XUVTwD0YSGAZJ0he:-KRUVDSrRLTFi3VvBIVy:boe1h0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Reading an expected number of errors (some will be nil) from the channel.
</details>
I'm going to leave it as is and ignore only the single expect error.

---

*[sql/create_test.go, line 229 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4pzeLyfpimAKI1yrQ:-KRURd3nM8js13bCaVBf:b-h3r5y6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4312c631cd422f3a466b4a22489b2792200bff/sql/create_test.go#L229)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I also don't think this insertion is necessary. And also I don't know if we should test that the ids are consecutive... Doesn't seem like a guarantee the database should be making.
</details>
I've added the test.  Seems like it holds up during stress.

---

*[sql/create_test.go, line 157 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUGg4wOP6MUTcjRJsy:-KRUSAwlk_zc6MOS_ZID:b-oz8tym) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L157)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> put a "parallel" in the test name
</details>
Good call.

---

*[sql/create_test.go, line 183 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUI-HbLBN2wRyfp7Zf:-KRUSiU3gAhKgHZU3psd:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L183)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> instead of a blanket `hideErrors`, pass in an expected error string.
</details>
Good call, done.

---

*[sql/create_test.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qSN59EihqSeIvJg8:-KRUUxcLn2FTyp10kt2u:bzfisqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L186)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> not done?
</details>
Was done, just not pushed.

---

*[sql/create_test.go, line 197 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUHA9eeib_2BC13yt2:-KRUUzwhBPVtyidlAvE3:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L197)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> could *not* be created
> What's this `t.Logf`? Haven't seen it before, just use `log.Infof()`? Or actually shouldn't this be `t.Errorf()` ?
</details>
fixed.

---

*[sql/create_test.go, line 236 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR4qfYMV8IKO6UJKV4Y:-KRURmZDy7Ms6jTtvk7a:bwiy867) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L236)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> btw we got rid of the deadline on the distsender context, so results should now be different
</details>
Interesting, I'll rebase and test it again. 

Looks like things are vastly improved.  I'll keep pushing the boundaries and see what shakes out.

---

*[sql/create_test.go, line 249 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUKFTJ9yyXmsXdmqd4:-KRURwSj96dLGC55A2r9:bggk0g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L249)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the bug that triggered these tests would only be observed by this test, which actually has failing `CREATE`s, right? So you're using the duplica name just to force some errors, you're not actually interested in testing anything about duplicate names, are you? If you, there's a way to inject stmt errors which I think should be useful here.
</details>
The duplicate test was just because it felt like a nice inclusion.  There wasn't a failing case for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Updated.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRUiFtmvCok580BdyKS:b-bw8e14)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 3 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 206 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV57IrAgUh5bZAqrOz:-KRV57IrAgUh5bZAqrP-:bm73bd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L206)):*
> ```Go
> 		t.Fatalf("expected %d tables created, only got %d", expectedNumOfTables, numTablesCreated)
> 	}
> 	descIDEnd := descIDStart + int64(numTablesCreated)
> ```

are you checking whether [keys.DescIDGenerator] is this value?

---

*[sql/create_test.go, line 220 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV3qmDZnTAOweFnEKi:-KRV3qmDZnTAOweFnEKj:b-cx950e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L220)):*
> ```Go
> 		if _, err := db.Exec(insertTableSQL, id); err != nil {
> 			t.Fatal(err)
> 		}
> ```

I don't see a strong need to inserting a value. Checking that the descriptors are there is enough IMHO.

---

*[sql/create_test.go, line 235 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV429iIMPDUmT_uG58:-KRV429iIMPDUmT_uG59:bi8kif9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L235)):*
> ```Go
> 			t.Fatalf("duplicate ID used for both tables %s and %s", tableName, otherTableName)
> 		}
> 		usedTableIDs[tableDesc.ID] = tableName
> ```

perhaps you could add all the ids into a slice, sort it, and check that the values increment from the min to the max value. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 3 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 249 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUKFTJ9yyXmsXdmqd4:-KRVMCH-NdNEiFc52meO:b1h8upy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L249)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The duplicate test was just because it felt like a nice inclusion.  There wasn't a failing case for it.
</details>
can you please remind me... The bug we found had to do with an error being swallowed, right? What was that error? Was it a "retry txn"? Or the infamous "context deadline exceeded"? I'm trying to figure out if either of these tests would have caught that bug.

---

*[sql/create_test.go, line 220 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV3qmDZnTAOweFnEKi:-KRVPQugELd1ygUirHos:bpx08g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L220)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I don't see a strong need to inserting a value. Checking that the descriptors are there is enough IMHO.
</details>
+1. In fact I don't see a need to even test the descriptors.

---

*[sql/create_test.go, line 235 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV429iIMPDUmT_uG58:-KRVPRdo8BIzfQ9FB_Wo:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L235)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> perhaps you could add all the ids into a slice, sort it, and check that the values increment from the min to the max value. 
</details>
+1

---

*[sql/create_test.go, line 291 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRVNNQgOG7hQiWnNN7W:-KRVNNQgOG7hQiWnNN7X:blfgvqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L291)):*
> ```Go
> // requests with same name are only filled once. This is the same test as
> // TestParallelCreateTables but in this test the tables names are all the same.
> func TestParallelCreateConflictingTables(t *testing.T) {
> ```

I don't know about this test... If it's not related to the error-swallowing bug, then it seems to me like an arbitrary race to test, and doesn't seem to be adding much to the other test. Or otherwise put, it's not clear to me what suspected race would be. If there would be one particular race we had in mind, we could probably set it up better than just blasting these CREATES.  Personally, I would get rid of it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRVPth6N3al-DZhcgO8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 3 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 291 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRVNNQgOG7hQiWnNN7W:-KRVd-trGC_VEBzkDByF:b-5639tu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L291)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I don't know about this test... If it's not related to the error-swallowing bug, then it seems to me like an arbitrary race to test, and doesn't seem to be adding much to the other test. Or otherwise put, it's not clear to me what suspected race would be. If there would be one particular race we had in mind, we could probably set it up better than just blasting these CREATES.  Personally, I would get rid of it.
</details>
I'll remove this test then.  I added it as an interesting case at the end of my original test and it has become it's own unneeded test.  And with that, I'll rearrange some of the functions that won't be needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Bram

I'd keep the other test to test the "If not exists" clause. No where do we
test that if we're creating many tables using the same name with the if not
exists clause one of them actually creates a table, rather than all failing
silently with no table created. Thanks

On Mon, Sep 12, 2016, 7:04 PM Bram Gruneir <notifications@github.com> wrote:

> Review status: 1 of 3 files reviewed at latest revision, 18 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *sql/create_test.go, line 291 at r7
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRVNNQgOG7hQiWnNN7W:-KRVd-trGC_VEBzkDByF:b-5639tu>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L291>):*
> *Previously, andreimatei (Andrei Matei) wrote…*
>
> I don't know about this test... If it's not related to the
> error-swallowing bug, then it seems to me like an arbitrary race to test,
> and doesn't seem to be adding much to the other test. Or otherwise put,
> it's not clear to me what suspected race would be. If there would be one
> particular race we had in mind, we could probably set it up better than
> just blasting these CREATES. Personally, I would get rid of it.
>
>
> I'll remove this test then. I added it as an interesting case at the end
> of my original test and it has become it's own unneeded test. And with
> that, I'll rearrange some of the functions that won't be needed.
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9182>*
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9182#issuecomment-246523315>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBAXtxgfGSQxZJUjootrlP00hsF13ks5qpdqBgaJpZM4J3HfK>
> .
>

BramGruneir commented:
Sure, that's an easy change to make.

---

Review status: 1 of 3 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[sql/create_test.go, line 249 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRUKFTJ9yyXmsXdmqd4:-KRZ8kq04EgPvCwud-Hd:b-z4d6z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c48df151fbef2f5f33bca652b63feb91fabb7862/sql/create_test.go#L249)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you please remind me... The bug we found had to do with an error being swallowed, right? What was that error? Was it a "retry txn"? Or the infamous "context deadline exceeded"? I'm trying to figure out if either of these tests would have caught that bug.
</details>
It was a retry txn that was not retrying.

---

*[sql/create_test.go, line 206 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV57IrAgUh5bZAqrOz:-KRZdLfIkWWJp8bUjImz:b-v5egdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L206)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> are you checking whether [keys.DescIDGenerator] is this value?
</details>
I was not.  Added.

---

*[sql/create_test.go, line 220 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV3qmDZnTAOweFnEKi:-KRZ9CUT85rVJTZfQ2gz:bwrh56u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L220)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> +1. In fact I don't see a need to even test the descriptors.
</details>
Actually, we need the insert or a show table or some sql of some sort or the testCluster just won't shut down.

---

*[sql/create_test.go, line 235 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRV429iIMPDUmT_uG58:-KRZ90_Vx-6p1ZRPq2De:baieiy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d606887c91a2cc86554a379b49848751e2a1f107/sql/create_test.go#L235)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> +1
</details>
I actually think that would take longer than doing what I've already got here.  We know the total number of tables, we know the range, and I check for duplicates and make sure all numbers are within the range.  So effectively the equivalent but no need to sort.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRZU6CMqcuU4Q6IgFlN:b-87dkkf)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I think this is good to go.  Can I get one final pass please @vivekmenezes?

---

Review status: 1 of 4 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRlHZMcbfTZlOmEF_VC:b-3ej3f3)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 4 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 182 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR_6SGQ5eKWT0go8ZtK:-KR_6SGQ5eKWT0go8ZtL:b72onsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/023dd7e65c1390860981e7fff891ceed230c76ad/sql/create_test.go#L182)):*
> ```Go
> 			val INT
> 		)`, id)
> 	if _, err := db.Exec(tableSQL); err != nil {
> ```

Given that you are forcing contention here I think you should retry this before this test starts failing and you need to skip it :-)
See runSQLWithRetry()

---

*[sql/create_test.go, line 219 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRn-J8UsauuZ_vKhctp:-KRn-J8UsauuZ_vKhctq:bcefz4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f70a17df7a9d2c47d2bc78119a2a7d85f99b3af/sql/create_test.go#L219)):*
> ```Go
> 	}
> 
> 	numTablesCreated := len(usedTableIDs)
> ```

limit the scope of numTablesCreated to the if stmt

---

*[storage/replica.go, line 1565 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRn0vOoVWxVW6dCmfeS:-KRn0vOoVWxVW6dCmfeT:b-ykbiu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f70a17df7a9d2c47d2bc78119a2a7d85f99b3af/storage/replica.go#L1565)):*
> ```Go
> 				if tryAbandon() {
> 					pErr = roachpb.NewError(&roachpb.NodeUnavailableError{})
> 				}
> ```

} else {
					log.Warningf(ctx, "unable to cancel expired Raft command %s", ba)
				}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRn1Nx8UkhIb8-mlvV3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 4 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 182 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR_6SGQ5eKWT0go8ZtK:-KRnkgS5LWPCXcPwdSD5:b-htk6s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/023dd7e65c1390860981e7fff891ceed230c76ad/sql/create_test.go#L182)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Given that you are forcing contention here I think you should retry this before this test starts failing and you need to skip it :-)
> See runSQLWithRetry()
</details>
Well this single statement should be retried automatically by the server on every retryable sql error.
That `IsSQLRetryError` function has `(connection reset by peer|connection refused|failed to send RPC|EOF)")`. None of these look retryable to me; I don't think they're supposed to happen in this test (and god knows what other tests do to cause this).
The function used to also guard for the infamous `context  deadline exceeded`, which did indeed have a connection to contention. But that was resolved and the error went away. In fact, we probably should've resolved the cause in the first place instead of papering over it :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 4 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 182 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR_6SGQ5eKWT0go8ZtK:-KRnoh3Grj5rvZXCwP9u:b-208y6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/023dd7e65c1390860981e7fff891ceed230c76ad/sql/create_test.go#L182)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Well this single statement should be retried automatically by the server on every retryable sql error.
> That `IsSQLRetryError` function has `(connection reset by peer|connection refused|failed to send RPC|EOF)")`. None of these look retryable to me; I don't think they're supposed to happen in this test (and god knows what other tests do to cause this).
> The function used to also guard for the infamous `context  deadline exceeded`, which did indeed have a connection to contention. But that was resolved and the error went away. In fact, we probably should've resolved the cause in the first place instead of papering over it :)
</details>
https://github.com/cockroachdb/cockroach/commit/0a61994b49acb7c1bfd239a09ff3b3d5ee4ce9c6 removed context deadline exceeded. Hmmm, its time to deprecate this function. @bram no need for it here. I'll remove it in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 4 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 182 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR_6SGQ5eKWT0go8ZtK:-KRo-YFBTUh7taSSuXpx:b-xqq90) ([raw file](https://github.com/cockroachdb/cockroach/blob/023dd7e65c1390860981e7fff891ceed230c76ad/sql/create_test.go#L182)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> https://github.com/cockroachdb/cockroach/commit/0a61994b49acb7c1bfd239a09ff3b3d5ee4ce9c6 removed context deadline exceeded. Hmmm, its time to deprecate this function. @bram no need for it here. I'll remove it in a separate PR.
</details>
https://github.com/cockroachdb/cockroach/pull/9426

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on storage/replica.go:
```diff
@@ -1559,6 +1559,10 @@ func (r *Replica) addWriteCmd(
 				} else {
 					log.Warningf(ctx, "unable to cancel expired Raft command %s", ba)
 				}
+			case <-r.store.stopper.ShouldQuiesce():
```
This will actually have the request spin hot on this code path if `tryAbandon` returns false, but if that happens the command should be executing and finish momentarily. I think we can risk this for now (or you would have to follow that same pattern as with `ctxDone` above).

tschottdorf replied:
I'm extracting this into a separate PR since we're seeing more failures related to this.

BramGruneir commented:
Go for it.  I'm distracted by family right now.

On Sep 16, 2016 16:28, "Tobias Schottdorf" <notifications@github.com> wrote:

> *@tschottdorf* commented on this pull request.
> ------------------------------
>
> In storage/replica.go <https://github.com/cockroachdb/cockroach/pull/9182>
> :
>
> > @@ -1559,6 +1559,10 @@ func (r *Replica) addWriteCmd(
>  				} else {
>  					log.Warningf(ctx, "unable to cancel expired Raft command %s", ba)
>  				}
> +			case <-r.store.stopper.ShouldQuiesce():
>
> I'm extracting this into a separate PR since we're seeing more failures
> related to this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9182>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABihubAspYl-1FcdxHlWKue2dPyUYfmTks5qqvv2gaJpZM4J3HfK>
> .
>

tschottdorf commented:
Done. You should be able to just drop that commit.

On Fri, Sep 16, 2016 at 4:44 PM Bram Gruneir <notifications@github.com>
wrote:

> Go for it. I'm distracted by family right now.
>
> On Sep 16, 2016 16:28, "Tobias Schottdorf" <notifications@github.com>
> wrote:
>
> > *@tschottdorf* commented on this pull request.
> > ------------------------------
> >
> > In storage/replica.go <
> https://github.com/cockroachdb/cockroach/pull/9182>
> > :
> >
> > > @@ -1559,6 +1559,10 @@ func (r *Replica) addWriteCmd(
> > } else {
> > log.Warningf(ctx, "unable to cancel expired Raft command %s", ba)
> > }
> > + case <-r.store.stopper.ShouldQuiesce():
> >
> > I'm extracting this into a separate PR since we're seeing more failures
> > related to this.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/cockroachdb/cockroach/pull/9182>, or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABihubAspYl-1FcdxHlWKue2dPyUYfmTks5qqvv2gaJpZM4J3HfK
> >
> > .
> >
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9182#issuecomment-247704433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135A-Ccb3YJ6Ef-z21ZYp5lrHPrpeaks5qqv99gaJpZM4J3HfK>
> .
>
-- 

-- Tobias

BramGruneir commented:
Cool.  Dropped.

---

Review status: 1 of 3 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[sql/create_test.go, line 182 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KR_6SGQ5eKWT0go8ZtK:-KRorVWUUew2ftaDVFze:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/023dd7e65c1390860981e7fff891ceed230c76ad/sql/create_test.go#L182)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> https://github.com/cockroachdb/cockroach/pull/9426
</details>
Done.

---

*[sql/create_test.go, line 219 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRn-J8UsauuZ_vKhctp:-KRorKolkikR5s0wqVhR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f70a17df7a9d2c47d2bc78119a2a7d85f99b3af/sql/create_test.go#L219)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> limit the scope of numTablesCreated to the if stmt
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-:-KRosnyNPuL95r8v3zc2:bfwt7ri)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5, 1 of 1 files at r6, 1 of 1 files at r9, 1 of 3 files at r15, 1 of 1 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[sql/create_test.go, line 187 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182#-KRt1kS46SJg1-eat-8q:-KRt1kS5ugWx3dI1xQxa:b-4349n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/152ec0de430d545037bd15553203067189df6251/sql/create_test.go#L187)):*
> ```Go
> 
> 	if _, err := db.Exec(tableSQL); err != nil {
> 		t.Fatalf("table %d: could not be created: %s", id, err)
> ```

bad practice to do this outside the main goroutine. again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9182)*
<!-- Sent from Reviewable.io -->
